### PR TITLE
895 move enabled disabled property from jobs to edges

### PR DESCRIPTION
**Pull request by benedictus-dev**
## Related issue
This PR addresses issue #895, focusing on the migration of the 'enabled/disabled' property from the 'jobs' to the 'edges' within our workflow management system
Fixes #

1. Introduces an 'enabled' boolean to edges, allowing for a more granular control over workflow 
2. Removes the 'enabled/disabled' toggles from the job panel, aligning with our strategy to separate the execution of jobs from decision-making processes.
## Review checklist

- [x] I have performed a **self-review** of my code
- [x] I have verified that all appropriate **authorization policies** have been implemented and tested
- [x] If needed, I have updated the **changelog**
- [x] Product has **QA'd** this feature

### CODE REVIEW

File: lib/lightning/workflows/edge.ex [hunk 1]
```diff
@@ -39,6 +40,8 @@ defmodule Lightning.Workflows.Edge do
 
     field :condition, Ecto.Enum, values: @conditions
 
+    field :enabled, :boolean
```
Comment (taylordowns2000) [inline lib/lightning/workflows/edge.ex:43]: i think this should default to true to keep parity with the old `job.ex`

File: lib/lightning_web/live/workflow_live/edit.ex [hunk 2]
```diff
@@ -682,6 +682,7 @@ defmodule LightningWeb.WorkflowLive.Edit do
 
   @impl true
   def handle_info({"form_changed", %{"workflow" => params}}, socket) do
+    IO.puts("=====================When to trigger me ===============")
```
Comment (taylordowns2000) [inline lib/lightning_web/live/workflow_live/edit.ex:685]: remove IO.inspect?

File: lib/lightning_web/live/workflow_live/workflow_params.ex [hunk 3]
```diff
@@ -13,7 +13,13 @@ defmodule LightningWeb.WorkflowNewLive.WorkflowParams do
     Map.merge(current_params, form_params, fn _key, current, next ->
       case {current, next} do
         {current, next} when is_list(current) and is_map(next) ->
-          merge_list_params(current, next)
+          if Map.has_key?(form_params, "edges") &&
```
Comment (taylordowns2000) [inline lib/lightning_web/live/workflow_live/workflow_params.ex:16]: i think this is the code that's setting downstream edges. please ensure that if someone disbales an edge, only that edge gets disabled.

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
hey @benedictus-dev , looking good! please address all of my inline change requests. 

the big thing (workflow_params, line 16) is that upon disabling an edge all downstream edges are getting disabled. it looks super deliberate so i'm guessing this was simply a miscommunication somewhere. please make sure that when a user disables an edge, only that edge gets disabled. other elements of the workflow should be left alone.

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
1. please re-write these migrations to use a single migration that _moves_ the boolean from jobs to edges. this is fairly tricky. see details below.
2. please add tests (chat @midigofrank and @elias-ba on slack and rebase on release-0.10.0 once they've finished fixing the workflow params test.)

details on the migration:
- currently the enabled/disabled status lives on the jobs table.
- we need to determine what edge is pointing at each job, and _move_ the value of that boolean to the leading edge.
- the exception here is that no edges which lead from a trigger to a job may be disabled.

so to say it another way:
- first add the enabled column to edges
- then set the value with this logic:
  - if the edge has a `source_trigger_id`, set `edge.enabled = true`
  - if the edge does not have a source trigger, set the `edge.enabled` value to the `jobs.enabled` value for the `edge.target_job_id`
- then finally drop the `jobs.enabled` column

File: test/lightning/workflows/query_test.exs [hunk 4]
```diff
@@ -60,21 +61,36 @@ defmodule Lightning.Workflows.QueryTest do
...
 
       _disabled_cronjob =
-        insert(:job, enabled: false, workflow: trigger.workflow)
+        insert(:job, workflow: trigger.workflow)
+
+      insert(:edge, %{
+        source_trigger: trigger,
+        target_job: _disabled_cronjob,
+        workflow: _disabled_cronjob.workflow,
+        enabled: false
```
Comment (taylordowns2000) [inline test/lightning/workflows/query_test.exs:80]: this is impossible, i think!

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
this works _great_ from a UX perspetive! nice job @benedictus-dev ! todo:

1. ~mark_for_deletion - is this doing what needs to be done? i thought we discussed setting the trigger to disabled.~ done
2. canonical_project - i see you've removed `enabled` from jobs. should i be added to edges?
3. `test/lightning/projects_test.exs` - i think this is related to number item 1. i think we discussed checking to see that triggers were disabled, rather than jobs.
4. `test/lightning/workflows/query_test.exs` - i'm not convinced that this enabled_cron_jobs_by_edge query is being tested properly with your changes. i'm also not sure it's being used anymore! how do we find a list of enabled cron workflows to create new attempts in the system?
5. see "impossible" comment on trigger-to-job edge with `enabled: false`. we should add validation here to stop this.
6. please remove remaining commented out code

File: test/lightning/workflows/query_test.exs [hunk 5]
```diff
@@ -60,21 +61,36 @@ defmodule Lightning.Workflows.QueryTest do
...
       webhook_trigger = insert(:trigger, type: :webhook)
 
       _non_cron_job =
-        insert(:job, enabled: true, workflow: webhook_trigger.workflow)
+        insert(:job, workflow: webhook_trigger.workflow)
+
+      insert(:edge, %{
+        source_trigger: webhook_trigger,
+        target_job: _non_cron_job,
+        workflow: _non_cron_job.workflow,
```
Comment (stuartc) [inline test/lightning/workflows/query_test.exs:91]: Variables prefixed with an underscore are a hint that the variable isn't supposed to be used: https://hexdocs.pm/elixir/main/naming-conventions.html#underscore-_foo

It was left there as a 'comment' of sorts, in tests they are permissible but we try and not have these kinds of variable names in the application code.

So in this situation, could you remove the `_` prefix from the variable.

File: lib/lightning/projects.ex [hunk 6]
```diff
@@ -465,11 +465,11 @@ defmodule Lightning.Projects do
       end
 
     Repo.transaction(fn ->
-      jobs = project_jobs_query(project) |> Repo.all()
+      workflows = project_workflows_query(project) |> Repo.all()
```
Comment (taylordowns2000) [inline lib/lightning/projects.ex:468]: this should be `triggers`

File: lib/lightning/workflows/query.ex [hunk 7]
```diff
@@ -36,7 +36,7 @@ defmodule Lightning.Workflows.Query do
     from(e in Edge,
       join: j in assoc(e, :target_job),
       join: t in assoc(e, :source_trigger),
-      where: t.type == :cron and t.enabled,
+      where: t.type == :cron and t.enabled and e.enabled,
```
Comment (taylordowns2000) [inline lib/lightning/workflows/query.ex:39]: remove `and e.enabled`

File: lib/lightning/projects.ex [hunk 8]
```diff
@@ -465,11 +465,11 @@ defmodule Lightning.Projects do
...
 
-      jobs
-      |> Enum.each(fn job ->
-        Lightning.Jobs.update_job(job, %{"enabled" => false})
+      workflows
+      |> Enum.each(fn workflow ->
+        Lightning.Workflows.mark_for_deletion(workflow)
```
Comment (taylordowns2000) [inline lib/lightning/projects.ex:472]: with this PR, you've moved `:enabled` from Jobs to Edges. On this line you should go from:
```
Lightning.Jobs.update_job(job, %{"enabled" => false})
```
to:
```
Lightning.Triggers.update_trigger(trigger, %{"enabled" => false})
```
There may not be an `update_trigger` function and you may need to make one or call `Repo.update()`

File: test/lightning/setup_utils_test.exs [hunk 9]
```diff
@@ -219,7 +217,7 @@ defmodule Lightning.SetupUtilsTest do
                });
              """
 
-      assert job_3.enabled
+      # assert job_3.enabled
```
Comment (taylordowns2000) [inline test/lightning/setup_utils_test.exs:220]: remove commented code

File: test/lightning/setup_utils_test.exs [hunk 10]
```diff
@@ -538,7 +534,7 @@ defmodule Lightning.SetupUtilsTest do
              fn(state => state);
              """
 
-      assert notify_upload_successful.enabled
+      # assert notify_upload_successful.enabled
```
Comment (taylordowns2000) [inline test/lightning/setup_utils_test.exs:537]: remove commented code

File: test/lightning/workflows_test.exs [hunk 11]
```diff
@@ -128,7 +130,6 @@ defmodule Lightning.WorkflowsTest do
 
       # Disabled Job
```
Comment (taylordowns2000) [inline test/lightning/workflows_test.exs:131]: rename this comment; jobs can't be disabled

**Review by taylordowns2000 — CHANGES_REQUESTED** [review]
looking really good @benedictus-dev . please address each of my final change requests and re-request a review. let's get this over the line!!!

**Review by taylordowns2000 — APPROVED** [review]

**Pull request merged**